airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+    
+    def set_up_primary_key_last_value_seen(self, primary_key):
+      primary_key_last_value_seen = {}
+      for key in primary_key:
+        primary_key_last_value_seen[key] = STARTING_PRIMARY_KEY_VALUE
+      return primary_key_last_value_seen
+
+    def __init__(self, db_connection: pyodbc.Cursor, table_name, stream, config):
+      self.db_connection = db_connection
+      self.table_name = table_name
+      self.properties = self.get_primary_key_from_airbyte_stream(stream) # self.get_properties_from_stream(stream)
+      self.primary_key_column = self.get_primary_key_from_airbyte_stream(stream)
+      self.incremental_column = self.get_incremental_column_from_airbyte_stream(stream)
+      self.primary_key_last_value_seen = self.set_up_primary_key_last_value_seen(self.primary_key_column)
+      self.incremental_most_recent_value_seen = None
+      self.json_schema = stream.json_schema
+      self.config = config # used for restoring the connection when it expires
+
+    def read_records(
+      self,
+      sync_mode: SyncMode,
+      cursor_field: Optional[List[str]] = None,
+      stream_slice: Optional[Mapping[str, Any]] = None,
+      stream_state: Optional[Mapping[str, Any]] = None,
+    ) -> Iterable[StreamData]:
+      try:
+        self.process_stream_state(stream_state)
+        self.db_connection.execute(self.generate_query(sync_mode, stream_slice))
+        number_records = 0
+        while True:
+          row = self.db_connection.fetchone()
+          if not row and number_records < NETSUITE_PAGINATION_INTERVAL:
+            break
+          # if number of rows >= page limit, we need to fetch another page
+          elif not row:
+            self.logger.info(f"Fetching another page for the stream {self.table_name}.  Current state is (primary key: {str(self.primary_key_last_value_seen)}, date: {self.incremental_most_recent_value_seen}")
+            number_records = 0
+            self.db_connection.execute(self.generate_query(sync_mode, stream_slice))
+            continue
+          # data from netsuite does not include columns, so we need to assign each value to the correct column name
+          serialized_data = self.serialize_row_to_response(row)
+          # before we yield record, update state
+          self.update_last_values_seen(serialized_data, sync_mode)
+          self.find_most_recent_date(serialized_data)
+          number_records = number_records + 1
+          #yield response
+          yield serialized_data
+      except Exception as e:
+        message = str(e)
+        if NETSUITE_CONNECTION_EXPIRED_FAILURE in message:
+          self.logger.info(f"Connection expired for stream {self.table_name}.  Reconnecting...")
+          cursor_constructor = NetsuiteODBCCursorConstructor()
+          new_db_connection = cursor_constructor.create_database_connection(self.config)
+          self.db_connection = new_db_connection
+          current_stream_state = self.get_updated_state(stream_state, {})
```

Comment:
I think this will break if stream_state is `None` when [trying to check for the presence of an item in current_stream_state](https://github.com/airbytehq/airbyte/pull/32968/files#diff-ec237bc9a560edd56705d285e834a64adfe4cc092ff3356a74349d2f2215adc2R173)

Reply:
Fixed!